src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -179,18 +191,40 @@ private void EnqueueParseSnapshotTask(Document newDocument)
                 // we will enqueue new one soon, cancel pending refresh right away
                 _reportChangeCancellationSource.Cancel();
 
-                var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-                var snapshot = newText.FindCorrespondingEditorTextSnapshot();
-                if (snapshot == null)
+                ITextSnapshot snapshot;
+                if (document != null)
                 {
-                    // It's possible that we're seeing a notification for an update that happened
-                    // just before the file was opened, and so the document we're given is still the
-                    // old one.
-                    return;
+                    var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    snapshot = newText.FindCorrespondingEditorTextSnapshot();
+                    if (snapshot == null)
+                    {
+                        // It's possible that we're seeing a notification for an update that happened
+                        // just before the file was opened, and so the document we're given is still the
+                        // old one.
+                        return;
+                    }
+
+                    // preemptively parse file in background so that when we are called from tagger from UI thread, we have tree ready.
+                    var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // This buffer isn't being tracked by the workspace. In Visual Studio, this
+                    // can happen if it is not in the Running Document Table. We'll try to do
+                    // basic syntax classification anyway.
+
+                    snapshot = _subjectBuffer.CurrentSnapshot;
+
+                    var languageName = _contentTypesToLanguageNames.FirstOrDefault(x => _subjectBuffer.ContentType.MatchesAny(x.Metadata.DefaultContentType))?.Metadata.Language;
+                    if (languageName != null)
```

Comment:
If `languageName` is `null`, you'll set `snapshot`, but not `document`, which will then flow into `_lastParsedDocument`.  Is that okay?